projects/graph/src/graph.py

```diff
@@ -3,7 +3,124 @@
 """
 
 
+class Vertex:
+    def __init__(self, value):
+        self.node = value
+        self.edges = set()
+
+
 class Graph:
     """Represent a graph as a dictionary of vertices mapping labels to edges."""
+
     def __init__(self):
-        pass  # TODO
+        self.vertices = {}
+
+    def add_vertex(self, value):
+        if value not in self.vertices:
+            vertex = Vertex(value)
+            self.vertices[vertex.node] = vertex.edges
+            return True
+        else:
+            return False
+
+    def add_edge(self, v1, v2):
+        if v1 in self.vertices and v2 in self.vertices:
+            self.vertices[v1].add(v2)
+            self.vertices[v2].add(v1)
+            return True
+        else:
+            if v1 not in self.vertices:
+                raise IndexError(
+                    f"Vertex {v1} is nonexistent!"
+                )  # IndexError more specific than Exception
+            elif v2 not in self.vertices:
+                raise IndexError(f"Vertex {v2} is nonexistent!")
+
+    def add_directed_edge(self, v1, v2):
+        if v1 in self.vertices and v2 in self.vertices:
+            self.vertices[v1].add(v2)
+            return True
+        else:
+            if v1 not in self.vertices:
+                raise IndexError(
+                    f"Vertex {v1} is nonexistent!"
+                )  # IndexError more specific than Exception
+            elif v2 not in self.vertices:
+                raise IndexError(f"Vertex {v2} is nonexistent!")
+
+    def bft(self, target):
+        q = []
+        visited = []
+        q.append(int(target))
+
+        if str(target) not in self.vertices:
+            raise IndexError(f"Vertex {target} is nonexistent!")
+
+        while len(q) > 0:
+            discovered = q.pop(0)
+            if discovered not in visited:
+                visited.append(discovered)
+                for c in self.vertices[str(discovered)]:
+                    q.append(int(c))
+
+        print(f"vertices: {self.vertices}")
+        return f"visited: {visited}"
+
+    def dft(self, target):
+        s = []
+        visited = []
+        s.append(int(target))
+
+        if str(target) not in self.vertices:
+            raise IndexError(f"Vertex {target} is nonexistent!")
+
+        while len(s) > 0:
+            discovered = s.pop()
+            if discovered not in visited:
+                visited.append(discovered)
+                for c in self.vertices[str(discovered)]:
```

Comment:
Is this targeting the edges of each child vertex or is it simply targeting every vertex as they appear in the dict? I'm curious if there are test cases related to the ordering of the vertices (in the dict and when generating edges) that would break this method as is. I will run some test when I get a chance. Having said that the code looks great!

Reply:
I am using the sets from the dictionaries and can see why it works.

After talking about it in code review and seeing Brady inserting the dictionary value as Vertex(value) this morning, I realized my implementation is different.

I do see the difference in what you're pointing out targeting off the edges in the class instead of off the edges in the dictionary. Much appreciated!